streaming/stream_session.cc

```diff
@@ -111,24 +112,30 @@ void stream_manager::init_messaging_service_handler(abort_source& as) {
             return make_ready_future<>();
         });
     });
-    ms.register_stream_mutation_fragments([this, &as] (const rpc::client_info& cinfo, streaming::plan_id plan_id, table_schema_version schema_id, table_id cf_id, uint64_t estimated_partitions, rpc::optional<stream_reason> reason_opt, rpc::source<frozen_mutation_fragment, rpc::optional<stream_mutation_fragments_cmd>> source) {
+    ms.register_stream_mutation_fragments([this, &as] (const rpc::client_info& cinfo, streaming::plan_id plan_id, table_schema_version schema_id, table_id cf_id, uint64_t estimated_partitions,
+            rpc::optional<stream_reason> reason_opt,
+            rpc::optional<service::session_id> session,
+            rpc::source<frozen_mutation_fragment, rpc::optional<stream_mutation_fragments_cmd>> source) {
         auto from = netw::messaging_service::get_source(cinfo);
         auto reason = reason_opt ? *reason_opt: stream_reason::unspecified;
+        service::frozen_topology_guard topo_guard = session.value_or(service::session_id());
```

Comment:
I thought that `frozen_topology_guard` is something to be used on wire, but the series uses almost everywhere, but on wire where simple uuid is transferred.

Reply:
frozen_topology_guard is equivalent to uuid in this solution, so we also pass it on wire.

It's fine to pass it around the node too, since it's cheaper than the materialized guard. We need to materialize the full guard around operations which have side-effects, or if we wanted to check for abort early. In the plumbing layers it's unnecessary.

Reply:
sessions are implementation detail of the guards, which I wanted to hide from topology guard users so that they are not coupled with a particular solution.

We need a distinction between frozen and regular guard, regardless of how you call it. "frozen" doesn't keep the topology coordinator on hold while it's alive, and it can be copied between shards and nodes by value, like frozen_mutation. Regular guard cannot be copied, and it will hold the topology coordinator.